### PR TITLE
bug: `BytesUtils.compare()` Arithmetic Overflow

### PR DESCRIPTION
**Pull request by adraffy**
The difference `int256 - int256` exceeds `int256` range.

Example: `type(int256).min - 1 < type(int256).min`

---

* change `BytesUtils.compare()` to fix overflow
* add test case for maximum signed difference (caused error before)
* to test: `bun run test test/dnssec-oracle/TestSolidityTests.ts`

### CODE REVIEW

File: contracts/utils/BytesUtils.sol [hunk 1]
```diff
@@ -77,11 +77,11 @@ library BytesUtils {
...
-    /*
-     * @dev Returns a positive number if `other` comes lexicographically after
-     *      `self`, a negative number if it comes before, or zero if the
-     *      contents of the two bytes are equal. Comparison is done per-rune,
-     *      on unicode codepoints.
+    /**
+     * @dev +1 if `other` comes lexicographically after `self`,
+     *      -1 if `other` comes before,
+     *      or 0 if the contents of the two bytes are equal.
+     *      Comparison is done bytewise.
```
Comment (Arachnid) [inline contracts/utils/BytesUtils.sol]: This should stay as it was - we don't want to unduly constrain implementations.

File: contracts/utils/BytesUtils.sol [hunk 2]
```diff
@@ -123,21 +123,25 @@ library BytesUtils {
                 b := mload(otherptr)
             }
             if (a != b) {
-                // Mask out irrelevant bytes and check again
-                uint256 mask;
-                if (shortest - idx >= 32) {
-                    mask = type(uint256).max;
-                } else {
-                    mask = ~(2 ** (8 * (idx + 32 - shortest)) - 1);
+                uint256 rest = shortest - idx;
```
Comment (Arachnid) [inline contracts/utils/BytesUtils.sol:126]: Can you leave the comment intact?

File: contracts/utils/BytesUtils.sol [hunk 3]
```diff
@@ -123,21 +123,25 @@ library BytesUtils {
...
+                if (rest < 32) {
+                    rest = (32 - rest) << 3; // bits to drop
+                    a >>= rest; // drop them
+                    b >>= rest;
                 }
-                int256 diff = int256(a & mask) - int256(b & mask);
-                if (diff != 0) return diff;
+                if (a < b) return -1;
+                if (a > b) return 1;
```
Comment (Arachnid) [inline contracts/utils/BytesUtils.sol]: Please put all if clauses in braces.

File: contracts/utils/BytesUtils.sol [hunk 4]
```diff
@@ -123,21 +123,25 @@ library BytesUtils {
...
             otherptr += 32;
         }
 
-        return int256(len) - int256(otherlen);
+        return
+            len == otherlen
+                ? int256(0)
+                : len < otherlen
+                    ? int256(-1)
+                    : int256(1);
```
Comment (Arachnid) [inline contracts/utils/BytesUtils.sol]: If we don't change the docstring, this change isn't needed either.

**Review by Arachnid — CHANGES_REQUESTED** [review]

**Review by Arachnid — APPROVED** [review]
